src/coprocessor/codec/chunk/chunk.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME(shirly): remove following later
+#![allow(dead_code)]
+use super::Result;
+use super::column::Column;
+use coprocessor::codec::Datum;
+use std::sync::Arc;
+use tipb::expression::FieldType;
+
+/// Chunk stores multiple rows of data in Apache Arrow format.
+/// See https://arrow.apache.org/docs/memory_layout.html
+/// Values are appended in compact format and can be directly accessed without decoding.
+/// When the chunk is done processing, we can reuse the allocated memory by resetting it.
+pub struct Chunk {
+    columns: Vec<Column>,
+}
+
+impl Chunk {
+    ///new creates a new chunk with field types and capacity.
+    pub fn new(tps: &[FieldType], cap: usize) -> Chunk {
+        let mut columns = Vec::with_capacity(tps.len());
+        for tp in tps {
+            columns.push(Column::new(tp, cap));
+        }
+        Chunk { columns }
+    }
+
+    // reset resets the chunk, so the memory it allocated can be reused.
+    // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
+    pub fn reset(&mut self) {
+        for column in &mut self.columns {
+            column.reset();
+        }
+    }
+
+    // num_cols returns the number of rows in the chunk.
+    pub fn num_cols(&self) -> usize {
+        self.columns.len()
+    }
+
+    // num_rows returns the number of rows in the chunk.
+    pub fn num_rows(&self) -> usize {
+        if self.columns.is_empty() {
+            0
+        } else {
+            self.columns[0].length()
+        }
+    }
+
+    pub fn append_datum(&mut self, col_idx: usize, v: &Datum) -> Result<()> {
+        self.columns[col_idx].append_datum(v)
+    }
+}
+
+struct ArcChunk {
```

Comment:
It seems wired to have `ArcChunk`, because we don't have things like `ArcRow`. Also it is not reasonable that a `Chunk` cannot get a row but an `Atomic Reference Counted Chunk` can. Consider embedding `Arc` into `Chunk`?

Reply:
no, `ArcChunk` is designed only for read.